src/Workspaces/Core/Portable/Shared/Extensions/ContextQuery/AbstractSyntaxContext.cs

```diff
@@ -80,21 +82,27 @@ private ISet<INamedTypeSymbol> ComputeOuterTypes(CancellationToken cancellationT
                 var containingType = enclosingSymbol.GetContainingTypeOrThis();
                 if (containingType != null)
                 {
-                    return containingType.GetContainingTypes().ToSet();
+                    return new Tuple<ISet<ITypeSymbol>, ITypeSymbol>(containingType.GetContainingTypesAndThis().ToSet(), containingType);
                 }
             }
 
-            return SpecializedCollections.EmptySet<INamedTypeSymbol>();
+            return new Tuple<ISet<ITypeSymbol>, ITypeSymbol>(SpecializedCollections.EmptySet<ITypeSymbol>(), null);
         }
 
-        public ISet<INamedTypeSymbol> GetOuterTypes(CancellationToken cancellationToken)
+        public IEnumerable<ITypeSymbol> GetOuterTypes(bool includeContainingType, CancellationToken cancellationToken)
         {
-            if (_outerTypes == null)
+            if (_outerTypesIncludingContainingTypeAndContainingType == null)
             {
-                Interlocked.CompareExchange(ref _outerTypes, ComputeOuterTypes(cancellationToken), null);
+                Interlocked.CompareExchange(ref _outerTypesIncludingContainingTypeAndContainingType, ComputeOuterTypes(cancellationToken), null);
+            }
+            var outerTypesIncludingContainingType = _outerTypesIncludingContainingTypeAndContainingType.Item1;
+            var containingType = _outerTypesIncludingContainingTypeAndContainingType.Item2;
+            if (includeContainingType)
+            {
+                return outerTypesIncludingContainingType;
             }
 
-            return _outerTypes;
+            return outerTypesIncludingContainingType.Where(x => x != containingType);
```

Comment:
It's possible for this to throw since `ComputeOuterTypes(...)` can return `Tuple<ISet<ITypeSymbol>, ITypeSymbol>(SpecializedCollections.EmptySet<ITypeSymbol>(), null)` in the case where there's no containing type.